python/test/backend/test_device_backend.py

```diff
@@ -0,0 +1,262 @@
+        if cache_path is None:
+            with tempfile.TemporaryDirectory() as tmpdir:
+                src = self._generate_launcher(constants, signature)
+                src_path = os.path.join(tmpdir, "main.c")
+                with open(src_path, "w") as f:
+                    f.write(src)
+                so = build_for_backend(name, src_path, tmpdir)
+                with open(so, "rb") as f:
+                    so_path = so_cache_manager.put(f.read(), so_name, binary=True)
+                    type(self).stub_so_path = so_path
+                    return so_path
+        else:
+            type(self).stub_so_path = cache_path
+            return cache_path
+
+    def _generate_launcher(self, constants, signature):
+        # generate glue code
+        src = """
+        #define __EXTENSION_BACKEND__
+        #include <Python.h>
+        #include <stdio.h>
+
+        static PyObject* launch_counter(PyObject* self, PyObject* args) {
+        static int64_t launch_counter = 0;
+        launch_counter += 1;
+        return PyLong_FromLong(launch_counter);
+        }
+
+        static PyObject* launch(PyObject* self, PyObject* args) {
+        if (PyErr_Occurred()) {
+            return NULL;
+        }
+        launch_counter(self, args);
+        // return None
+        Py_INCREF(Py_None);
+        return Py_None;
+        }
+
+        static PyMethodDef ModuleMethods[] = {
+        {"launch", launch, METH_VARARGS, "Entry point for all kernels with this signature"},
+        {"launch_counter", launch_counter, METH_VARARGS, "Entry point to get launch counter"},
+        {NULL, NULL, 0, NULL} // sentinel
+        };
+
+        static struct PyModuleDef ModuleDef = {
+        PyModuleDef_HEAD_INIT,
+        \"__triton_launcher\",
+        NULL, //documentation
+        -1, //size
+        ModuleMethods
+        };
+
+        PyMODINIT_FUNC PyInit___triton_launcher(void) {
+        PyObject *m = PyModule_Create(&ModuleDef);
+        if(m == NULL) {
+            return NULL;
+        }
+        PyModule_AddFunctions(m, ModuleMethods);
+        return m;
+        }
+        """
+
+        return src
+
+
+def test_thirdparty_backend():
+    register_backend("cpu", ExtensionBackend)
+
+    @triton.jit
+    def kernel(in_ptr0, out_ptr0, xnumel, XBLOCK: tl.constexpr):
+        xnumel = 10
+        xoffset = tl.program_id(0) * XBLOCK
+        xindex = xoffset + tl.arange(0, XBLOCK)[:]
+        xmask = xindex < xnumel
+        x0 = xindex
+        tmp0 = tl.load(in_ptr0 + (x0), xmask)
+        tl.store(out_ptr0 + (x0 + tl.zeros([XBLOCK], tl.int32)), tmp0, xmask)
+
+    inp = torch.randn(10)
+    out = torch.randn(10)
+    kernel[(10,)](inp, out, 10, XBLOCK=16)
+    spec = importlib.util.spec_from_file_location("__triton_launcher", ExtensionBackend.stub_so_path)
+    mod = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(mod)
+    launch_counter = getattr(mod, "launch_counter")
```

Comment:
feels like this should be unifed into a function that is understandble similar to register backend.

Reply:
@micmelesse , it is for test purposes to count if the launcher has been invoked as the backend is a dummy backend and do nothing internally.